contracts/external/cw-token-swap/src/types.rs

```diff
@@ -0,0 +1,399 @@
+                                        .map(|a| deps.api.addr_validate(&a).unwrap().to_string()),
+                                    code_id,
+                                    msg,
+                                    funds: amount_to_pay,
+                                    label,
+                                }
+                                .into())
+                            }
+                        })
+                        .collect::<Result<Vec<CosmosMsg>, ContractError>>()?;
+
+                    // Make sure that the funds we try to send, matches exactly to the total amount swapped.
+                    if total_send_funds != total_amount {
+                        return Err(ContractError::WrongFundsCalculation {});
+                    }
+
+                    Ok(cosmos_msgs)
+                } else {
+                    Err(ContractError::InvalidSendMsg {})
+                }
+            }
+        }
+    }
+}
+
+/// This function check the given funds are valid and adds the amount to the total funds.
+/// returns the new total.
+pub fn add_to_total(
+    total_funds: Uint128,
+    funds: &[Coin],
+    denom: &str,
+) -> Result<Uint128, ContractError> {
+    let amount = must_pay(
+        &MessageInfo {
+            sender: Addr::unchecked(""),
+            funds: funds.to_owned(),
+        },
+        denom,
+    )?;
+
+    Ok(total_funds + amount)
+}
+
+/// Information about a counterparty in this escrow transaction and
+/// their promised funds.
+#[cw_serde]
+pub struct Counterparty {
+    /// The address of the counterparty.
+    pub address: String,
+    /// The funds they have promised to provide.
+    pub promise: TokenInfo,
+}
+
+impl Counterparty {
+    pub fn into_checked(self, deps: Deps) -> Result<CheckedCounterparty, ContractError> {
+        Ok(CheckedCounterparty {
+            address: deps.api.addr_validate(&self.address)?,
+            provided: false,
+            promise: self.promise.into_checked(deps)?,
+            send_msg: None,
+        })
+    }
+}
+
+#[cw_serde]
+pub struct CheckedCounterparty {
+    pub address: Addr,
+    pub promise: CheckedTokenInfo,
+    pub provided: bool,
+    pub send_msg: Option<Vec<CosmosMsg>>,
+}
+
+impl CheckedCounterparty {
+    pub fn add_send_msg(
+        &mut self,
+        deps: Deps,
+        send_msg: Option<SendMessage>,
+    ) -> Result<(), ContractError> {
+        self.send_msg = match send_msg {
+            Some(msg) => Some(msg.into_checked_cosmos_msgs(deps, self.promise.clone())?),
+            None => None,
+        };
+        Ok(())
+    }
+    pub fn remove_send_msg(&mut self) {
+        self.send_msg = None;
+    }
+}
+
+/// Information about the token being used on one side of the escrow.
+#[cw_serde]
+pub enum TokenInfo {
+    /// A native token.
+    Native { denom: String, amount: Uint128 },
+    /// A cw20 token.
+    Cw20 {
+        contract_addr: String,
+        amount: Uint128,
+    },
+}
```

Comment:
How about renaming this to `SwapInfo` and taking the messages to be executed as fields.

```rust
pub enum SwapInfo {
    /// A native token.
    Native { denom: String, amount: Uint128, on_completion: Vec<NativeSendMsg> },
    /// A cw20 token.
    Cw20 {
        contract_addr: String,
        amount: Uint128,
        on_completion: Vec<Cw20SendMsgs>
    },
}
```

This gives us two advantages:

1. When creating a swap agreement, all of the terms are known at instantiation time (as opposed to providing messages during funding).
2. We don't need `ContractError::InvalidSendMsg` as we get correctness by construction here.

The checked version of this could also contain the exact `CosmosMsg` to return, as yielded by `into_checked_cosmos_msgs `.